pysnark/fixedpoint.py

```diff
@@ -28,107 +28,171 @@
+        return "{" + str(self.val()) + "}"
     
     @classmethod
     def _ensurefxp(cls, val):
-        if isinstance(val,LinCombFxp):
+        """
+        Converts an integer, float, or LinComb into the internal fixed point representation
+        and converts the input into a LinCombFxp
+        """
+        if isinstance(val, LinCombFxp):
             return val
-        elif isinstance(val,LinComb):
-            return cls.fromvar(val, True)
-        else:
-            v = _tofxpval(val, True)
-            return cls(LinComb.ONE*v)
+        if isinstance(val, LinComb):
+            val = LinCombFxp.add_scaling(val)
+            return LinCombFxp(val)
+        if isinstance(val, int) or isinstance(val, float):
+            val = LinCombFxp.add_scaling(val)
+            return LinCombFxp(ConstVal(val))
+        raise RuntimeError("Wrong type for _ensurefxp")
     
     def __add__(self, other):
-        other = self._ensurefxp(other)
-        return LinCombFxp(self.lc+other.lc)
-        
+        """
+        Adds a LinCombFxp with an integer, float, LinComb, or another LinCombFxp
+        Costs 0 constraints
+        """
+        if isinstance(other, int) or isinstance(other, float) or isinstance(other, LinComb):
+            other = LinCombFxp.add_scaling(other)
+            return LinCombFxp(self.lc + other)
+        if isinstance(other, LinCombFxp):
+            return LinCombFxp(self.lc + other.lc)
+        return NotImplemented
+
     __radd__ = __add__
     
     def __sub__(self, other):
-        return self+(-other)
+        return self + (-other)
     
     def __rsub__(self, other):
-        return other+(-self)    
+        return other + (-self)
 
     def __mul__(self, other):
-        if is_base_value(other):
-            return LinCombFxp(self.lc*other)
-        
-        other = self._ensurefxp(other)
-        if other is NotImplemented: return NotImplemented
-        
-        ret = LinCombFxp(PrivVal((self.lc.value*other.lc.value+(1<<(LinCombFxp.res-1)))>>LinCombFxp.res))
-            
-        # mul error: should be in [-2^f,2^f]
-        diff = (1<<LinCombFxp.res)*ret.lc-self.lc*other.lc
-        (diff + (1<<LinCombFxp.res)).assert_positive(LinCombFxp.res+1)
-        ((1<<LinCombFxp.res) - diff).assert_positive(LinCombFxp.res+1)
-
-        return ret        
-    
+        """
+        Multiplies a LinCombFxp with an integer, float, LinComb, or another LinCombFxp
+        Costs 0 constraints to multiply with an integer or a float
+        Costs 1 constraint to multiply with a LinComb or LinCombFxp
+        """
+        if isinstance(other, int):
+            return LinCombFxp(self.lc * other)
+        if isinstance(other, float):
+            other = LinCombFxp.add_scaling(other)
+            return LinCombFxp((self.lc * other) / (1 << resolution))
+        if isinstance(other, LinComb):
+            other = LinCombFxp.add_scaling(other)
+            return LinCombFxp((self.lc * other) / (1 << resolution))

```

Comment:
This is a multiplication and a right-shift/division by 2**res. The right-shift uses a bit decomposition, so it will be costly (costing bit_length constraints). This could be avoided by simply doing other = LinCombFxp(self.lc*other), not sure why the original code did not do that....

Reply:
Thanks, I'll fix this.